src/yarpmanager/src-manager/clusterWidget.cpp

```diff
@@ -0,0 +1,477 @@
+}
+
+void ClusterWidget::onKillSelected()
+{
+    QList<QTreeWidgetItem*> selectedItems = ui->nodestreeWidget->selectedItems();
+    foreach (QTreeWidgetItem *it, selectedItems)
+    {
+        int itr = it->text(5).toInt();
+        ClusNode node = cluster.nodes[itr];
+        if (!node.onOff)
+        {
+            continue;
+        }
+
+        string cmdKillYarprun = getSSHCmd(node.user, node.name, node.ssh_options);
+
+        cmdKillYarprun = cmdKillYarprun + " killall -9 yarprun";
+
+        if (system(cmdKillYarprun.c_str()) != 0)
+        {
+            std::string err = "ClusterWidget: failed to kill yarprun on" + node.name;
+            logError(QString(err.c_str()));
+        }
+    }
+    yarp::os::Time::delay(2.0);
+    onCheckAll();
+}
+
+void ClusterWidget::onExecute()
+{
+    if (ui->lineEditExecute->text().trimmed().size() == 0)
+    {
+        return;
+    }
+
+    QList<QTreeWidgetItem*> selectedItems = ui->nodestreeWidget->selectedItems();
+    foreach (QTreeWidgetItem *it, selectedItems)
+    {
+        int itr = it->text(5).toInt();
+        ClusNode node = cluster.nodes[itr];
+
+        string cmdExecute = getSSHCmd(node.user, node.name, node.ssh_options);
+
+        cmdExecute = cmdExecute + " "+ ui->lineEditExecute->text().toStdString();
+
+        if (system(cmdExecute.c_str()) != 0)
+        {
+            std::string err = "ClusterWidget: failed to run "+ ui->lineEditExecute->text().toStdString() + " on " + node.name;
+            logError(QString(err.c_str()));
+        }
+    }
+}
+
+void ClusterWidget::onNodeSelectionChanged()
+{
+    if(ui->nodestreeWidget->selectedItems().isEmpty())
+    {
+        ui->runSelBtn->setDisabled(true);
+        ui->stopSelBtn->setDisabled(true);
+        ui->killSelBtn->setDisabled(true);
+        ui->executeBtn->setDisabled(true);
+    }
+    else
+    {
+        ui->runSelBtn->setDisabled(false);
+        ui->stopSelBtn->setDisabled(false);
+        ui->killSelBtn->setDisabled(false);
+        ui->executeBtn->setDisabled(false);
+    }
+}
+
+
+
+void ClusterWidget::addRow(string name, string display, string user,
+                           bool onOff, bool log, int id)
+{
+    QStringList stringList;
+    stringList <<""<< QString(name.c_str()) << QString(display.c_str()) << QString(user.c_str())<< "" <<QString(std::to_string(id).c_str());
+    QTreeWidgetItem* it = new QTreeWidgetItem(stringList);
+    ui->nodestreeWidget->addTopLevelItem(it);
+    ui->nodestreeWidget->setItemWidget((QTreeWidgetItem *) it, 4, new QCheckBox(this));
+
+    //initialize checkboxes
+    qobject_cast<QCheckBox*>(ui->nodestreeWidget->itemWidget((QTreeWidgetItem *)it, 4))->setChecked(log);
+
+    //initialize icon
+    if (onOff)
+    {
+        it->setIcon(0, QIcon(":/computer-available22.svg"));
+    }
+    else
+    {
+        it->setIcon(0, QIcon(":/computer-unavailable22.svg"));
+    }
+
+}
+
+std::string ClusterWidget::getSSHCmd(std::string user, std::string host, std::string ssh_options)
```

Comment:
This kind of warnings is good... instead of `std::string host` you should pass a `const std::string& host`. This saves an expansive copy of the string. Same for the other similar warnings.